src/nvim/edit.c

```diff
@@ -3846,42 +3829,68 @@ static bool ins_bs(int c, int mode, int *inserted_space_p)
     // Handle deleting one 'shiftwidth' or 'softtabstop'.
     if (mode == BACKSPACE_CHAR
         && ((p_sta && in_indent)
-            || ((get_sts_value() != 0
-                 || tabstop_count(curbuf->b_p_vsts_array))
+            || ((get_sts_value() != 0 || tabstop_count(curbuf->b_p_vsts_array))
                 && curwin->w_cursor.col > 0
                 && (*(get_cursor_pos_ptr() - 1) == TAB
                     || (*(get_cursor_pos_ptr() - 1) == ' '
                         && (!*inserted_space_p || arrow_used)))))) {
-      colnr_T vcol;
-      colnr_T want_vcol;
-
       *inserted_space_p = false;
-      // Compute the virtual column where we want to be.  Since
-      // 'showbreak' may get in the way, need to get the last column of
-      // the previous character.
-      getvcol(curwin, &curwin->w_cursor, &vcol, NULL, NULL);
-      colnr_T start_vcol = vcol;
-      dec_cursor();
-      getvcol(curwin, &curwin->w_cursor, NULL, NULL, &want_vcol);
-      inc_cursor();
+
+      bool const use_ts = !curwin->w_p_list || curwin->w_p_lcs_chars.tab1;
+      int const lnum = curwin->w_cursor.lnum;
+      char *const line = ml_get_buf(curbuf, lnum);
+      char *const end_ptr = line + curwin->w_cursor.col;
+
+      colnr_T vcol = 0;
+      colnr_T space_vcol = 0;
+      StrCharInfo sci = utf_ptr2StrCharInfo(line);
+      StrCharInfo space_sci = sci;
+      bool prev_space = false;
+      do {
+        bool cur_space = ascii_isspace(sci.chr.value);
```

Comment:
`ascii_iswhite()` should be used, not `ascii_isspace()`.
```suggestion
        bool cur_space = ascii_iswhite(sci.chr.value);
```